foursight_core/react/api/react_routes.py

```diff
@@ -0,0 +1,279 @@
+    response is returned, otherwise we go ahead with the function/route invocation.
+    A request is authorized iff it is AUTHENTICATED, i.e. the user has successfully
+    logged in, AND also has PERMISSION to access the specified environment. If the
+    request is not authorized an HTTP 401 is returned; if the request is authenticated
+    but is not authorized (for the specified environment) and HTTP 403 is returned.
+
+    The info to determine this is pass via the authtoken cookie, which is a (server-side)
+    JWT-signed-encode value containing authentication info and list allowed environment
+    for the user; this value/cookie is set server-side at login time.
+
+    Note that ONLY two React API routes should NOT be authorization protected by
+    this decorator: the /{environ}/header and the /{environ}/logout endpoints/routes.
+    """
+    def wrapper(*args, **kwargs):
+        if not kwargs or len(kwargs) < 1:
+            raise Exception("Invalid arguments to requires_authorization decorator!")
+        env = kwargs["environ"]
+        request = app.current_request.to_dict()
+        authorize_response = app.core.react_authorize(request, env)
+        if not authorize_response or not authorize_response["authorized"]:
+            response = app.core.create_standard_response("route_requires_authorization")
+            response.body = authorize_response
+            # HTTP 401 - Unauthorized (more precisely: Unauthenticated):
+            # Request has no or invalid credentials.
+            # HTTP 403 - Forbidden (more precisely: Unauthorized):
+            # Request has valid credentials but no privileges for resource.
+            if not authorize_response["authenticated"]:
+                response.status_code = HTTP_UNAUTHENTICATED
+            else:
+                response.status_code = HTTP_UNAUTHORIZED
+            return response
+        return f(*args, **kwargs)
+    return wrapper
+
+
+class ReactRoutes:
+
+    def __init__(self):
+        super(ReactRoutes, self).__init__()
+
+    # ----------------------------------------------------------------------------------------------
+    # Foursight React API routes.
+    # ----------------------------------------------------------------------------------------------
+
+    @staticmethod
+    @app.route(ROUTE_PREFIX + "reactapi/{environ}/header", methods=["GET"], cors=CORS)
+    def reactapi_route_header(environ: str):
+        request = app.current_request.to_dict()
+        return app.core.reactapi_header(request=request, env=environ)
+
+    @staticmethod
+    @app.route(ROUTE_PREFIX + "reactapi/header", methods=["GET"], cors=CORS)
+    def reactapi_route_header_noenv():
+        request = app.current_request.to_dict()
+        return app.core.reactapi_header(request=request, env=app.core.get_default_env())
+
+    @staticmethod
+    @app.route(ROUTE_PREFIX + "reactapi/{environ}/logout", methods=["GET"], cors=CORS)
+    def reactapi_route_logout(environ: str):
```

Comment:
Probably not a big deal but shouldn't you need to be authenticated in order to logout? Can someone hit logout without being logged in and trigger any weird behavior?

Reply:
Good question, but I think it is harmless/fine. Calling this when not authenticated merely sends HTTP 302 to login page, and expires the authtoken cookie. Though Idunno, maybe for completeness/correctness if not logged in it should return an HTTP 401, i.e. should be protected; I can go either way.

Reply:
Yeah, a message that cleanly says "You are already logged out." is preferred. People who do this to make sure they are logged out shouldn't be batted into thinking they did something wrong. It never hurts to be prudent, but issuing an error (even though it's functionally equivalent) may make them feel bad and like they shouldn't try.

Reply:
I kept this /logout endpoint as unprotected (no need to be logged in, i.e. if not logged in it will not send HTTP 401), because I feel that - just in case (who knows) - we don't want to prevent a logout attempt under any circumstance (IMO). However, as Kent suggested, I did add a (JSON) body to the (HTTP 302) response says "Already logged out." if not logged in or "Logged out." if logged in.